### ISSUE TITLE
Be able to set the logging level for istio and envoy logs from Kiali UI

### ISSUE THREAD

**Issue by lucasponce** [kiali/kiali#1525]
This feature would add being able to set the logging level from inside the Kiali UI workload -> logs tab.
This would be a big help when trying to debug something in istio using the logs and not having to go back out to change the logging level. And especially useful for the envoy proxy log level (--proxyLogLevel=debug)

JIRA:
https://issues.jboss.org/browse/KIALI-3013

**Comment by lucasponce** [comment]
I'm not sure if this feature can overlaps with a regular k8s console.
I agree it would be a good idea to have some control, but not sure with all those imply: Kiali start managing Pods for restart operations, config changes, etc.

**Comment by lucasponce** [comment]
This sounds valid as soon as we are adding more features around EnvoyConfig.

Istioctl allows to change the proxy config, and in particular, to modify the log level of a particular pod.

```
istioctl proxyu-config log --level ...
```

As logs tab relies on proxy logic, it sounds interesting to explore if that feature could be added at Workload level.

I mean Workload, because this would mean a re-start of the pod.

In general, Workload management tasks are out of the scope of Kiali, but as it happens with istio labels, this specific action, to change the log level of a container sounds correct and in line with other similar operations related to the EnvoyConfig.

cc @xeviknal as I suspect that the implementation used around accessing to the EnvoyConfig could be similar to this.

I don't add it to the Project, as it may require more discussion, but this issue sounds ok from my side.

**Comment by lucasponce** [comment]
Related to https://github.com/kiali/kiali/discussions/4082 discussion

**Comment by nrfox** [comment]
> I mean Workload, because this would mean a re-start of the pod.

In testing this by `kubectl exec` into the proxy pod and doing `curl -XPOST localhost:15000/logging?level=debug`, the log level updated successfully without a pod restart. If the istio proxy works the same way then a pod restart shouldn't be necessary for updating this.

**Comment by nrfox** [comment]
How about having the dropdown just to the left of the "log lines" dropdown?

![loglevellocation](https://user-images.githubusercontent.com/6226732/130290915-f11acf3d-153d-459e-8f7c-7dc9c9ed774c.png)

**Comment by jshaughn** [comment]
Is it actually possible to know the current log level?  Or just to set a new one?  I'm not sure whether I'd like to add more to the toolbar or instead control it from the kebab.

**Comment by nrfox** [comment]
> Is it actually possible to know the current log level? Or just to set a new one? I'm not sure whether I'd like to add more to the toolbar or instead control it from the kebab.

Good question! Yes, you can list the current loggers by sending the same request without a query param `curl -XPOST http://localhost:15000/logging`. https://www.envoyproxy.io/docs/envoy/latest/operations/admin#post--logging

**Comment by nrfox** [comment]
Another thing to consider for this feature is changing the log level of a specific logger vs. changing the log level for every logger. From the implementation side I don't believe it would be much work to support both since it's just a matter of setting the correct query param but supporting specific loggers would probably change the UI and how these options are laid out.

**Comment by jshaughn** [comment]
To start I'd say all loggers and see whether anyone asks for, or submits, an enhancement for individual loggers.

**Comment by nrfox** [comment]
So digging into this a little more I think there's two different use cases here for changing the log level:

1. Adjusting the log level for a specific workload. Say the global log level is `warning` but for a specific workload(s) you want the log level to be `info`. Here you want every pod in that workload to have the same `info` log level. That would be pretty tedious to do with the proxy but Istio supports updating the log level for a workload through [resource annotations](https://istio.io/latest/docs/reference/config/annotations/). The `Deployment` for that workload can be updated with the resource annotation thereby updating the log level for every pod. This would trigger a rolling restart for each pod.
2. Adjusting the log level for a specific pod. You are looking at the logs for a specific pod and want to increase the log level from `info` to `debug` in order to debug some envoy or proxy related issue. This is where the proxy is required since updating a pod's annotation directly will not update the proxy until the pod is restarted. However once the pod is restarted the pod's annotations will be updated to match the `Deployment`'s pod template spec.

IMO this issue is covering the second scenario but just wanted to point that out. The first scenario would be covered by https://github.com/kiali/kiali/issues/4179.

**Comment by jshaughn** [comment]
This should only be concerned with changing log level without a pod restart.  A pod restart will change everything regarding to the workload detail and would require a refresh of the entire detail, after the pod(s) restart.   As mentioned above, I don't think we want Kiali to do that sort of non-Istio-config workload manipulation (unless community really demands it).

**Comment by nrfox** [comment]
Added an action to the kebab dropdown on the pod logs page:

![setloglevel](https://user-images.githubusercontent.com/6226732/131519063-2fca5678-9206-4981-8635-c3b7c1e63644.png)

A problem with this approach is that it doesn't offer the ability to select which log level to set. There would need to be a separate action for each log level. 

Instead, clicking the kebab action could open a modal window that let's you choose which log level to set? That seems a bit overkill. 

Another option could be to have an expanded list on hovering over the "set log level" kebab action. When the user's mouse hovers "set log level", a second dropdown list appears to the right of the kebab menu with the different log level options. This might be more work to implement than the first two but perhaps offers a better UX.

A third option is what is mentioned above: https://github.com/kiali/kiali/issues/1525#issuecomment-902944493 showing the current log level as a separate dropdown menu from the kebab. This would also require implementing a "get logging level" route on the backend.

If we don't need or want to show the current log level, I'm leaning towards option 2 of an expanded kebab action menu. @jshaughn wdyt?

**Comment by jshaughn** [comment]
> If we don't need or want to show the current log level, I'm leaning towards option 2 of an expanded kebab action menu. @jshaughn wdyt?

How about if the dropdown just showed something like:

```
collapse Toolbar
match via Regex
show Timestamps
set Log Level
----------------
  trace
  debug
  info
  warn
  error
```
And the current log level is grayed out

**Comment by nrfox** [comment]
> > If we don't need or want to show the current log level, I'm leaning towards option 2 of an expanded kebab action menu. @jshaughn wdyt?
> 
> How about if the dropdown just showed something like:
> 
> ```
> collapse Toolbar
> match via Regex
> show Timestamps
> set Log Level
> ----------------
>   trace
>   debug
>   info
>   warn
>   error
> ```
> 
> And the current log level is grayed out

I like that approach too.

Showing the "current log level" is a little tricky since there doesn't seem to be a "global" log level option for the proxy. Rather there are 20+ individual loggers returned from `localhost:15000/logging`:

```
istio-proxy@ratings-v1-b6994bb9-cwprv:/$ curl -XPOST localhost:15000/logging 
active loggers:
  admin: warning
  aws: warning
  assert: warning
  backtrace: warning
  cache_filter: warning
  client: warning
  config: warning
  connection: warning
  conn_handler: warning
  decompression: warning
  dubbo: warning
  envoy_bug: warning
  ext_authz: warning
  rocketmq: warning
  file: warning
  filter: warning
  forward_proxy: warning
  grpc: warning
  hc: warning
  health_checker: warning
  http: warning
  http2: warning
  hystrix: warning
  init: warning
  io: warning
  jwt: warning
  kafka: warning
  lua: warning
  main: warning
  matcher: warning
  misc: error
  mongo: warning
  quic: warning
  quic_stream: warning
  pool: warning
  rbac: warning
  redis: warning
  router: warning
  runtime: warning
  stats: warning
  secret: warning
  tap: warning
  testing: warning
  thrift: warning
  tracing: warning
  upstream: warning
  udp: warning
  wasm: warning
```

Each logger could have it's own log level. To determine the "current" log level, I was planning on tallying up the occurrences of each and going with the max. My guess is that most of the time you'd get back the log level you'd expect. This is primarily a debugging feature so my guess is that users will not look at this to determine what the log level is for their workload so much as they will use this to "turn on/off debug logging" for a particular proxy that they are looking at and the "current" log level is just helpful to know when setting the log level for the individual proxy.

**Comment by jshaughn** [comment]
@nrfox OK, if it doesn't really make sense I'm fine not showing a "current" log level, just setting is fine.  This will consolidate all loggers to the selected log level, right?

**Comment by nrfox** [comment]
> @nrfox OK, if it doesn't really make sense I'm fine not showing a "current" log level, just setting is fine. This will consolidate all loggers to the selected log level, right?

Yes they'll all be updated to the same level. It's possible to target individual loggers with the envoy admin interface but I didn't think that was necessary for this feature since this is a temporary way to adjust your log level for debugging and that would add more complexity on the front end.

**Comment by nrfox** [comment]
This was fixed by kiali/kiali#2236
